safe_url_check/lib/safe_url_check.dart

```diff
@@ -0,0 +1,182 @@
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+/// Utility library to check if an untrusted URL is broken or not.
+///
+/// This makes an HTTP `HEAD` request after checking that the hostname does not
+/// resolve to a private IPv4 address or local unique IPv6 address.
+///
+/// **Example**
+/// ```dart
+/// import 'package:safe_url_check/safe_url_check.dart';
+///
+/// Future<void> main() async {
+///   // Check if https://google.com is a broken URL.
+///   final exists = await safeUrlCheck(
+///     Uri.parse('https://google.com'),
+///     userAgent: 'myexample/1.0.0 (+https://example.com)',
+///   );
+///   if (exists) {
+///     print('The url: https://google.com is NOT broken');
+///   }
+/// }
+/// ```
+library safe_url_check;
+
+import 'dart:io';
+import 'dart:async';
+
+import 'package:retry/retry.dart';
+
+import 'src/private_ip.dart';
+import 'src/unique_local_ip.dart';
+import 'src/version.dart';
+
+const _defaultUserAgent = 'package:safe_url_check/$packageVersion '
+    '(+https://github.com/google/dart-neats/tree/master/safe_url_check)';
+
+/// Check if [url] is available, without allowing access to private networks.
+///
+/// This will make a `HEAD` request to [url] with given [userAgent], following
+/// redirects if the hostname of URLs do not resolve to a private IP address.
+///
+/// Each I/O is set to timeout after [timeout] duration. This coupled with the
+/// large number of retries allow means that this operation can be slow.
+///
+/// It is good practice to set a custom [userAgent], this allows servers to see
+/// which bots are requesting their resources. This package defaults to the
+/// following user-agent:
+/// ```
+/// User-Agent: package:safe_url_check/1.0.0 (+https://github.com/google/dart-neats/tree/master/safe_url_check)
+/// ```
+Future<bool> safeUrlCheck(
+  Uri url, {
+  int maxRedirects = 8,
+  String userAgent = _defaultUserAgent,
+  HttpClient client,
+  RetryOptions retryOptions = const RetryOptions(),
+  Duration timeout = const Duration(seconds: 90),
+}) async {
+  ArgumentError.checkNotNull(url, 'url');
+  ArgumentError.checkNotNull(maxRedirects, 'maxRedirects');
+  ArgumentError.checkNotNull(userAgent, 'userAgent');
+  ArgumentError.checkNotNull(retryOptions, 'retryOptions');
+  if (maxRedirects < 0) {
+    throw ArgumentError.value(
+      maxRedirects,
+      'maxRedirects',
+      'must be a positive integer',
+    );
+  }
+
+  try {
+    // Always call with a client
+    if (client == null) {
+      final c = HttpClient();
+      try {
+        return await _safeUrlCheck(
+          url,
+          maxRedirects,
+          c,
+          userAgent,
+          retryOptions,
+          timeout,
+        );
+      } finally {
+        c.close(force: true);
```

Comment:
Nit: So you only close if you make your own client.
That is not obviously correct to me.
Should the absense of closing be documented?

Consider restructuring so there is only one call to _safeUrlCheck, and the closing happens conditionally.